### PR TITLE
James 1664 : integration work with OpenPaaS Inbox project

### PR DESCRIPTION
**Pull request by aduprat**
The main difficulty when implementing this issue was to take care of the difference between properties and header properties:
- properties are case sensitive
- header properties are case insensitive

What do you think of introducing a new class HeaderProterty ?

### CODE REVIEW

File: server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/GetMessagesMethodTest.java [hunk 1]
```diff
@@ -283,4 +290,67 @@ public void getMessagesShouldReturnFilteredHeaderPropertyWhenAsked() throws Exce
...
+            .content(startsWith("[[\"messages\","))
+            .extract()
+            .asString();
+
+        String firstResponsePath = "$.[0].[1]";
+
+        assertThat(jsonPath.parse(response).<Integer>read("$.length()")).isEqualTo(1);
+        assertThat(jsonPath.parse(response).<Integer>read(firstResponsePath + ".list.length()")).isEqualTo(0);
+        assertThat(jsonPath.parse(response).<Integer>read(firstResponsePath + ".notFound.length()")).isEqualTo(1);
+    }
```
Comment (fvn-linagora) [inline server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/GetMessagesMethodTest.java]: would be more terse and readable using jsonPath assertions in then() clause. Like so:
             .body("$", hasSize(1))
            .body("[0][1].list", hasSize(0))
            .body("[0][1].notFound", hasSize(1));

File: server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/GetMessagesMethodTest.java [hunk 2]
```diff
@@ -283,4 +290,67 @@ public void getMessagesShouldReturnFilteredHeaderPropertyWhenAsked() throws Exce
...
+        .then()
+            .statusCode(200)
+            .content(startsWith("[[\"messages\","))
+            .extract()
+            .asString();
+
+        String firstResponsePath = "$.[0].[1]";
+        String firstMessagePath = firstResponsePath + ".list[0]";
+
+        assertThat(JsonPath.parse(response).<Integer>read("$.length()")).isEqualTo(1);
```
Comment (fvn-linagora) [inline server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/GetMessagesMethodTest.java]: same here:
        .then()
            .statusCode(200)
            .content(startsWith("[[\"messages\","))
            .body("$", hasSize(1))
            .body("[0][1].list", hasSize(1))
            .body("[0][1].list[0].id", equalTo("username@domain.tld|mailbox|1"))
            .body("[0][1].list[0].threadId", notNullValue())
            .body("[0][1].list[0].mailboxIds", not(isEmptyString()))
            .body("[0][1].list[0].subject", equalTo("my test subject"));

File: server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/GetMessagesMethodTest.java (hunk 1 above)
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/GetMessagesMethodTest.java]: Would be cool to drop jsonPath dependency. I don't know why I needed to switch but if stock rest-assured works, we should use it.

Comment (aduprat) [inline server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/GetMessagesMethodTest.java]: We have to find a better tool for that.
Then all jmap integration tests have to be rewritten, so i think it should be done in a different issue.

File: server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/GetMessagesMethodTest.java (hunk 2 above)
Comment (aduprat) [inline server/protocols/jmap-integration-testing/src/test/java/org/apache/james/jmap/methods/GetMessagesMethodTest.java]: same comment

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java [hunk 3]
```diff
@@ -90,73 +93,83 @@
...
-        if (input.contains(MessageProperty.body)) {
+
+    private Set<MessageProperty> ensureContainsMandatoryFields(Optional<ImmutableSet<MessageProperty>> requestedProperties) {
+        return MANDATORY_PROPERTIES.stream()
+            .filter(mandatoryProperty -> propertyToAdd(mandatoryProperty, requestedProperties))
+            .collect(Collectors.toImmutableSet());
+    }
+
+    private boolean propertyToAdd(MessageProperty property, Optional<ImmutableSet<MessageProperty>> requestedProperties) {
+        return requestedProperties.isPresent() &&
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: I propose : 
```
return requestedProperties.ifPresent(x -> x.contains(property)).orElse(false);
```
If I'm wrong with `x -> x.contains(property)` try to declare a Function with a name to give to `ifPresent`

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java [hunk 4]
```diff
@@ -90,73 +93,83 @@
...
         }
     }
-    
-    private void handleBody(Set<MessageProperty> input, Set<MessageProperty> toAdd, Set<MessageProperty> toRemove) {
-        if (input.contains(MessageProperty.body)) {
+
+    private Set<MessageProperty> ensureContainsMandatoryFields(Optional<ImmutableSet<MessageProperty>> requestedProperties) {
+        return MANDATORY_PROPERTIES.stream()
+            .filter(mandatoryProperty -> propertyToAdd(mandatoryProperty, requestedProperties))
+            .collect(Collectors.toImmutableSet());
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: why don't you return the stream here ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java [hunk 5]
```diff
@@ -90,73 +93,83 @@
...
+    private void handleHeadersProperties(Optional<ImmutableSet<MessageHeaderProperty>> headerProperties, Set<MessageProperty> toAdd, Set<MessageProperty> toRemove) {
+        if (headerProperties.isPresent() && !headerProperties.get().isEmpty()) {
+            toAdd.add(MessageProperty.headers);
         }
     }
-    
-    private void handleBody(Set<MessageProperty> input, Set<MessageProperty> toAdd, Set<MessageProperty> toRemove) {
-        if (input.contains(MessageProperty.body)) {
+
+    private Set<MessageProperty> ensureContainsMandatoryFields(Optional<ImmutableSet<MessageProperty>> requestedProperties) {
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: all this is overly complex, why don't you just write `Sets.difference(mandatory, request)`

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java [hunk 6]
```diff
@@ -90,73 +93,83 @@
         Preconditions.checkArgument(request instanceof GetMessagesRequest);
         GetMessagesRequest getMessagesRequest = (GetMessagesRequest) request;
         Optional<ImmutableSet<MessageProperty>> requestedProperties = getMessagesRequest.getProperties();
+        Optional<ImmutableSet<MessageHeaderProperty>> headerProperties = getMessagesRequest.getHeaderProperties();
         return Stream.of(JmapResponse.builder().clientId(clientId)
                             .response(getMessagesResponse(mailboxSession, getMessagesRequest, requestedProperties))
                             .responseName(RESPONSE_NAME)
-                            .properties(requestedProperties.map(this::handleSpecificProperties))
-                            .filterProvider(Optional.of(buildFilteringHeadersFilterProvider(requestedProperties)))
+                            .properties(handleSpecificProperties(requestedProperties, headerProperties))
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: this little change makes nested calls a mess, can't we manage this "isPresent()" more gracefully ?

File: server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java [hunk 7]
```diff
@@ -90,73 +93,83 @@
         Preconditions.checkArgument(request instanceof GetMessagesRequest);
         GetMessagesRequest getMessagesRequest = (GetMessagesRequest) request;
         Optional<ImmutableSet<MessageProperty>> requestedProperties = getMessagesRequest.getProperties();
+        Optional<ImmutableSet<MessageHeaderProperty>> headerProperties = getMessagesRequest.getHeaderProperties();
```
Comment (mbaechler) [inline server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMessagesMethod.java]: Request objects is supposed to match the specification and I guess HeaderProperties is not part of it.

**Comment by mbaechler** [comment]
I think you have some work to do to make everything much simpler than what you propose here. Using streams and collections don't forbid a bit of object encapsulation if it serves readability.

**Comment by aduprat** [comment]
Refactored in:
https://github.com/linagora/james-project/pull/115

**Pull request closed without merging**